### PR TITLE
[LMSclean] AIRCoP transformation Step 1 (spatial splitting)

### PR DESCRIPTION
**Pull request by feiwang3311**

### CODE REVIEW

File: src/main/scala/lms/transformation/distributedTensorAIRCoP.scala [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  // this is the gradient map from OLD value tensors to NEW gradient tensors
+  val grad_map = mutable.HashMap[Backend.Sym, TENSOR]()
+  val momentum_map = mutable.HashMap[Backend.Sym, TENSOR]()
+
+  def traverseModule(ns: Seq[Node], res: Block): Backend.Exp = {
+    // Step 1: Collection Phase
+    ns.foreach { n => aircopCollect(n, forwardNodes, weightNodes, backwardNodes, grad_map, momentum_map, transform) }
+    // result of the block
+    implicit val pos = Adapter.oldSourceMap(ns.last.n)
+    (() => {
```
Comment (Kraks) [inline src/main/scala/lms/transformation/distributedTensorAIRCoP.scala:37]: If you want to do something lazy, it seems you can write
```
lazy val assignGrad = { ... }
assignGrad +=: backwardNodes
```

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorBase.scala [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  }
+}
+
+
+trait FixedSizeDistributedTensorOpsBase extends Dsl {
+
+  import PrimitiveTypeLess._
+  import FixedSizeDistributedTensorTypeLess._
+
+  def module[T](f: () => Rep[Tensor[T]])(implicit __pos: SourceContext) = {
```
Comment (Kraks) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorBase.scala:158]: why the unit argument is necessary? it looks like `f` can be just passed by name `=> Rep[Tensor[T]]`

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  def Mul(x: TENSOR, y: TENSOR, anno: Anno = NAnno)(implicit __pos: SourceContext): TENSOR =
+    ElemWiseNoBroadCasting(x, y, anno, __pos)("tensor_mult")
+
+  def Div(x: TENSOR, y: TENSOR, anno: Anno = NAnno)(implicit __pos: SourceContext): TENSOR =
+    ElemWiseNoBroadCasting(x, y, anno, __pos)("tensor_div")
+
+  override def mergable_dims(node: Node) = node match {
+    case Node(s, op, tt::anno::(x:Backend.Sym)::(y:Backend.Sym)::_, _)
+        if (op == "tensor_add" || op == "tensor_minus" || op == "tensor_mult" || op == "tensor_div") =>
```
Comment (Kraks) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala:36]: perhaps making a separate function as the predicate?

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      val x_type = (new TENSOR(x, useOldMetadata=true)).tensor_type
+      val y_type = (new TENSOR(y, useOldMetadata=true)).tensor_type
+      (x_type.shape zip y_type.shape).toList map { case (a:Size, b:Size) => (a.dim, b.dim) }
```
Comment (Kraks) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala:39]: `toList` is not necessary here

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    case _ => super.mergable_dims(node)
+  }
+
+  override def aircopCollect(node: Node, forwardNodes: mutable.ArrayBuffer[Node],
+      weightNodes: mutable.ArrayBuffer[Node], backwardNodes: mutable.ArrayBuffer[()=>Unit],
+      grad_map: mutable.HashMap[Backend.Sym, TENSOR],
+      momentum_map: mutable.HashMap[Backend.Sym, TENSOR],
```
Comment (Kraks) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala:46]: Nit: `forwardNodes`/`backwardNodes` vs `grad_map`/`momentum_map`, better to use a unified naming style

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      transform: Backend.Exp => Backend.Exp) = node match {
+
+    case Node(s, "tensor_add", tt::Backend.Const(anno:Anno)::(a:Backend.Sym)::(b:Backend.Sym)::_, _) =>
+      implicit val pos = Adapter.oldSourceMap(s)
+      // save forward op in forwardNodes
+      forwardNodes += node
+      // save backward op in backwardNodes
+      (() => {
```
Comment (Kraks) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala:54]: why does it need to be thunked and then forced immediately?

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala (hunk 4 above)
Comment (feiwang3311) [reply src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala:39]: the x_type.shape is Seq, which is not what I need from this expression (List)

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala (hunk 5 above)
Comment (feiwang3311) [reply src/main/scala/lms/transformation/distributed_tensor/distributedTensorBinary.scala:46]: will fix

**Pull request merged**